blog/tests/test_views.py

```diff
@@ -0,0 +1,108 @@
+        self.future_post = Post.objects.create(author=self.user,
+                                               title='Title Test 4',
+                                               text='Text Test 123 !@#',
+                                               published_date=datetime(day=4, month=4, year=2222, tzinfo=self.tz))
+
+    def tearDown(self):
+        """Clean-up test data."""
+        del self.first_post
+        del self.second_post
+        del self.third_post
+        del self.future_post
+        del self.tz
+        del self.client
+        del self.user
+
+    def test_post_list(self):
+        """Test list of posts."""
+        response = self.client.get(reverse('post_list'))
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.context['posts']), 3)
+        self.assertTemplateUsed(response, 'blog/post_list.html', 'blog/base.html')
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=1, year=2000, tzinfo=self.tz)):
+            self.assertNotContains(response, self.future_post)
+
+    def test_sorting_post_list(self):
+        """Test post sorting."""
+        response = self.client.get(reverse('post_list'), {'order_by': 'published_date'})
+        posts = response.context['posts']
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=1, year=2018, tzinfo=self.tz)):
+            self.assertEqual(posts[2].published_date, datetime(day=1, month=1, year=2017, tzinfo=self.tz))
+            self.assertEqual(posts[1].published_date, datetime(day=3, month=3, year=2014, tzinfo=self.tz))
+            self.assertEqual(posts[0].published_date, datetime(day=2, month=2, year=2011, tzinfo=self.tz))
+            self.assertNotContains(response, self.future_post)
+
+    def test_post_detail_view(self):
+        """Test detail page of specific post."""
+        response = self.client.get(reverse('post_detail', kwargs={'pk': 9999}))
+        self.assertEqual(404, response.status_code)
+        response = self.client.get(reverse('post_detail', kwargs={'pk': self.first_post.pk}))
+        self.assertEqual(200, response.status_code)
+
+    def test_post_new_view(self):
+        """Test view for add post."""
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
+        self.assertTrue(authorization)
+        response = self.client.post(reverse('post_new'),
+                                    {'author': self.user,
+                                     'title': 'Test New Post',
+                                     'text': 'Test Text',
+                                     },
+                                    follow=True)
+        self.assertEqual(200, response.status_code)
+        self.assertContains(response, 'Test New Post')
+
+    def test_post_edit(self):
+        """Test view for edit post."""
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
```

Comment:
Test access without auth as well.

Reply:
?

Reply:
if I remove the authorization string, it will knock this error

Reply:
Which means you didn't implement raising HTTP 401 in case of unauthorized user.
Also, it's good to try editing another user's post.